File: lib/services/exporter.dart
import 'dart:typed_data';
import 'package:excel/excel.dart';
import '../models/nurse.dart';
import '../models/shift_type.dart';

class Exporter {
  static Future<Uint8List?> exportToExcel(
      List<Nurse> nurses, int year, int month) async {
    var excel = Excel.createExcel();
    // Bypassing a bug in the Excel package where renaming throws "Unsupported operation"
    var sheetName = excel.getDefaultSheet() ?? "Sheet1";
    var ws = excel[sheetName];

    // Headers
    List<dynamic> headers = ["No.", "Name"];
    for (int d = 1; d <= 31; d++) {
      headers.add(d.toString());
    }
    headers.addAll([
      "M Hours",
      "E Hours",
      "N Hours",
      "Total Hours"
    ]);
    ws.appendRow(headers);

    // Row Data
    for (int i = 0; i < nurses.length; i++) {
      var nurse = nurses[i];
      List<dynamic> row = [
        (i + 1),
        nurse.name
      ];

      for (int d = 1; d <= 31; d++) {
        var shift = nurse.getShift(d);
        if (shift != ShiftType.off) {
          row.add(shift.value);
        } else {
          row.add(""); // Empty for OFF
        }
      }

      int mHrs = nurse.countShiftType(ShiftType.morning) * 6;
      int eHrs = nurse.countShiftType(ShiftType.evening) * 6;
      int nHrs = nurse.countShiftType(ShiftType.night) * 12;

      row.addAll([
        mHrs,
        eHrs,
        nHrs,
        (mHrs + eHrs + nHrs)
      ]);
      ws.appendRow(row);
    }

    // Daily Totals
    List<dynamic> totalM = ["Total M", ""];
    List<dynamic> totalE = ["Total E", ""];
    List<dynamic> totalN = ["Total N", ""];

    for (int d = 1; d <= 31; d++) {
      int mCount = 0;
      int eCount = 0;
      int nCount = 0;
      for (var nurse in nurses) {
        var shift = nurse.getShift(d);
        if (shift == ShiftType.morning) mCount++;
        if (shift == ShiftType.evening) eCount++;
        if (shift == ShiftType.night) nCount++;
      }
      totalM.add(mCount);
      totalE.add(eCount);
      totalN.add(nCount);
    }

    ws.appendRow([]); // Empty row
    ws.appendRow(totalM);
    ws.appendRow(totalE);
    ws.appendRow(totalN);

    // Style the totals roughly (Excel package has limited styling unless Pro, but we can do bold)
    var boldStyle = CellStyle(bold: true);
    for (int r = ws.maxRows - 3; r < ws.maxRows; r++) {
      for (int c = 0; c < ws.maxCols; c++) {
        var cell = ws.cell(CellIndex.indexByColumnRow(columnIndex: c, rowIndex: r));
        cell.cellStyle = boldStyle;
      }
    }

    // Official Signatures
    ws.appendRow([]);
    ws.appendRow([]);
    ws.appendRow(["", "", "", "Date: ___________________"]);
    ws.appendRow([]);
    ws.appendRow(["", "", "", "Head nurse signature: ___________________"]);

    var fileBytes = excel.save();
    if (fileBytes != null) {
      return Uint8List.fromList(fileBytes);
    }
    return null;
  }
}

File: src/exporter.py
from openpyxl import Workbook
from openpyxl.styles import Font, Alignment, PatternFill
from typing import List, Any
from models import Nurse, ShiftType

class Exporter:
    @staticmethod
    def export_to_excel(nurses: List[Nurse], year: int, month: int, filename: str):
        wb = Workbook()
        ws = wb.active
        ws.title = f"Schedule {month}-{year}"
        
        # Headers
        headers = ["No.", "Name"] + [str(d) for d in range(1, 32)] + ["M Hours", "E Hours", "N Hours", "Total Hours"]
        ws.append(headers)
        
        # Row Data (Placeholder)
        for i, nurse in enumerate(nurses, 1):
            row = [i, nurse.name]
            # Days 1-31
            for d in range(1, 32):
                shift = nurse.get_shift(d)
                row.append(shift.value if shift != ShiftType.OFF else "")
            
            # Totals (as hours: count × shift_hours)
            m_hrs = nurse.count_shift_type(ShiftType.MORNING) * 6
            e_hrs = nurse.count_shift_type(ShiftType.EVENING) * 6
            n_hrs = nurse.count_shift_type(ShiftType.NIGHT) * 12
            row.append(m_hrs)
            row.append(e_hrs)
            row.append(n_hrs)
            row.append(m_hrs + e_hrs + n_hrs)
            
            ws.append(row)

        # --- Daily Totals ---
        # Calculate totals per column (day)
        total_m = ["Total M", ""]
        total_e = ["Total E", ""]
        total_n = ["Total N", ""]
        
        for d in range(1, 32):
            m_count = 0
            e_count = 0
            n_count = 0
            for nurse in nurses:
                shift = nurse.get_shift(d)
                if shift == ShiftType.MORNING: m_count += 1
                elif shift == ShiftType.EVENING: e_count += 1
                elif shift == ShiftType.NIGHT: n_count += 1
            
            total_m.append(m_count)
            total_e.append(e_count)
            total_n.append(n_count)
            
        ws.append([]) # Empty row
        ws.append(total_m)
        ws.append(total_e)
        ws.append(total_n)

        from openpyxl.styles import Font
        # Style the totals (optional but nice)
        for row_idx in range(ws.max_row - 2, ws.max_row + 1):
            for cell in ws[row_idx]:
                cell.font = Font(bold=True)

        wb.save(filename)
        print(f"Exported to {filename}")
